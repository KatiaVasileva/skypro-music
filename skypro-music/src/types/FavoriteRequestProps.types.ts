export type FavoriteRequestProps = {
  id: number | undefined;
  access: string;
  refresh: string;
};
