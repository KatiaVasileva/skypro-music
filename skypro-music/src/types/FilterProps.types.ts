export type FilterProps = {
  performers: Array<string>;
  genres: Array<string>;
};
