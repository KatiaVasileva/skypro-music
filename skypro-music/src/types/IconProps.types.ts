export type IconProps = {
  wrapperClass?: string;
  iconClass?: string;
  name: string;
  onClick?: (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => void;
};
