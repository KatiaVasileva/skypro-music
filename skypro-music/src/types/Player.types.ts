import { ChangeEventHandler } from "react";

export type ProgressBarProps = {
  max: number;
  value: number;
  step: number;
  onChange: ChangeEventHandler<HTMLInputElement>;
};
