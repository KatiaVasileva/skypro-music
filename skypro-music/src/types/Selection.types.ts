export type Selection = {
    _id: number;
    name: string;
    items: Array<number>;
    owner: Array<number>;
    __v: number;
};