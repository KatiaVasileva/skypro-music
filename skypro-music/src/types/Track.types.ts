import { Key } from "react";

export type Track = {
  _id: Key;
  name: string;
  author: string;
  album: string;
  duration_in_seconds: number;
  release_date: Date;
  genre: Array<string>;
};