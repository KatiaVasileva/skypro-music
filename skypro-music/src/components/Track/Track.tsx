"use client";

import { Track } from "@/types/Track.types";
import styles from "./Track.module.css";
import {
  setPlayingState,
  setPlaylistState,
  setTrackIndexState,
  setTrackState,
} from "@/store/features/trackSlice";
import { useAppDispatch, useAppSelector } from "@/store/store";
import Icon from "../Icon/Icon";
import { useLikeTrack } from "@/hooks/useLikeTracks";
import { formatTime } from "@/utils/helpers";

type TrackItemProps = {
  track: Track;
  tracks: Array<Track>;
};

function TrackItem({ track, tracks }: TrackItemProps) {
  const dispatch = useAppDispatch();
  const playlistState = useAppSelector((state) => state.track.playlistState);
  const myPlaylistState = useAppSelector(
    (state) => state.track.myPlaylistState
  );
  const playingState = useAppSelector((state) => state.track.playingState);
  const trackIndexState = useAppSelector(
    (state) => state.track.trackIndexState
  );
  const shuffleActiveState = useAppSelector(
    (state) => state.track.shuffleActiveState
  );
  const shuffledPlaylistState = useAppSelector(
    (state) => state.track.shuffledPlaylistState
  );
  const trackState = useAppSelector((state) => state.track.trackState);

  const { isLiked, handleLike } = useLikeTrack({ track });

  const handleTracks = () => {
    dispatch(setTrackState(track));
    dispatch(setPlaylistState({ tracks: tracks }));
    dispatch(setPlayingState(true));
    dispatch(
      setTrackIndexState(
        shuffleActiveState
          ? shuffledPlaylistState.indexOf(track)
          : playlistState.indexOf(track)
      )
    );
  };

  const handleLikeButton = async (event: React.MouseEvent<HTMLElement>) => {
    handleLike(event);
  };

  return (
    <div
      className={styles.playlistItem}
      key={track._id}
      onClick={() => {
        handleTracks();
      }}
    >
      <div className={styles.playlistTrack}>
        <div className={styles.trackTitle}>
          <div className={styles.imageContainer}>
            {!shuffleActiveState && trackState &&
              playlistState.indexOf(track) === trackIndexState && (
                <div
                  className={
                    playingState ? styles.playingDotAnimated : styles.playingDot
                  }
                ></div>
              )}
            {shuffleActiveState && trackState &&
              shuffledPlaylistState.indexOf(track) === trackIndexState && (
                <div
                  className={
                    playingState ? styles.playingDotAnimated : styles.playingDot
                  }
                ></div>
              )}

            <Icon
              wrapperClass={styles.trackTitleImage}
              iconClass={styles.trackTitleSvg}
              name="icon-note"
            />
          </div>
          <div>
            <a className={styles.trackTitleLink} href="#">
              {track.name} <span className={styles.trackTitleSpan}></span>
            </a>
          </div>
        </div>
        <div className={styles.trackAuthor}>
          <a className={styles.trackAuthorLink} href="http://">
            {track.author}
          </a>
        </div>
        <div className={styles.trackAlbum}>
          <a className={styles.trackAlbumLink} href="http://">
            {track.album}
          </a>
        </div>

        <Icon
          iconClass={isLiked ? styles.trackLikeSvgActive : styles.trackLikeSvg}
          name="icon-like"
          onClick={handleLikeButton}
        />

        <div>
          <span className={styles.trackTimeText}>
            {formatTime(track.duration_in_seconds)}
          </span>
        </div>
      </div>
    </div>
  );
}

export default TrackItem;
