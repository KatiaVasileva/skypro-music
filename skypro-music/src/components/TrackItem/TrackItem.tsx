"use client";

import styles from "./TrackItem.module.css";
import Icon from "../Icon/Icon";
import { formatTime } from "@/utils/helpers";
import { TrackItemProps } from "@/types/TrackItemProps.types";
import { useState } from "react";
import { useAppDispatch } from "@/store/store";
import { setPlayingState } from "@/store/features/playingSlice";
import { setTrackState } from "@/store/features/trackSlice";

type StatusProps = {
  isDotActive: boolean;
}

function Dot({isDotActive} : StatusProps) {
  return (
    <div className={isDotActive ? styles.playingDotAnimated : styles.playingDotInactive}></div> 
  );
}

function TrackItem({ track }: TrackItemProps) {
  //   const playingState = useAppSelector((state) => state.playing.playingState);
  const dispatch = useAppDispatch();

  const [isCurrentTrackPlaying, setIsCurrentTrackPlaying] = useState(false);
  const [trackIndex, setTrackIndex] = useState(0);

  // const onClickTrack = () => {
  //   dispatch(setTrackState(track));
  //   dispatch(setPlayingState(true));
  //   setTrackIndex(track._id);
  //   // setIsCurrentTrackPlaying(track._id === trackIndex);
  // };

  return (
    <div
      className={styles.playlistItem}
      key={track._id}
      onClick={() => {
        dispatch(setTrackState(track));
        dispatch(setPlayingState(true));
        setTrackIndex(track._id);
      }}
    >
      <div className={styles.playlistTrack}>
        <div className={styles.trackTitle}>
          <div className={styles.imageContainer}>
            {/* {track._id === trackIndex && (
              <div className={styles.playingDotAnimated}></div>
            )} */}
            <Dot isDotActive={trackIndex === track._id}/>
            <Icon
              wrapperClass={styles.trackTitleImage}
              iconClass={styles.trackTitleSvg}
              name="icon-note"
            />
          </div>
          <div>
            <a className={styles.trackTitleLink} href="#">
              {track.name} <span className={styles.trackTitleSpan}></span>
            </a>
          </div>
        </div>
        <div className={styles.trackAuthor}>
          <a className={styles.trackAuthorLink} href="http://">
            {track.author}
          </a>
        </div>
        <div className={styles.trackAlbum}>
          <a className={styles.trackAlbumLink} href="http://">
            {track.album}
          </a>
        </div>
        <Icon iconClass={styles.trackTimeSvg} name="icon-like" />
        <div>
          <span className={styles.trackTimeText}>
            {formatTime(track.duration_in_seconds)}
          </span>
        </div>
      </div>
    </div>
  );
}

export default TrackItem;
