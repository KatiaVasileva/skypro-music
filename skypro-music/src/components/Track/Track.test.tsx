import ReduxProvider from "../../store/ReduxProvider";
import { render } from "@testing-library/react";
import { Track } from "@/types/Track.types";
import TrackItem from "./Track";

describe("Track component", () => {
  it("renders correctly", () => {
    const track = {
      _id: 1,
      name: "Track1",
      author: "Author1",
      album: "Album1",
      duration_in_seconds: 100,
      release_date: "2024-09-22",
      genre: ["Rock"],
      track_file: "track1.mp3",
    };

    const tracks: Array<Track> = [
      {
        _id: 1,
        name: "Track1",
        author: "Author1",
        album: "Album1",
        duration_in_seconds: 100,
        release_date: "2024-09-22",
        genre: ["Rock"],
        track_file: "track1.mp3",
      },
    ];

    const { asFragment } = render(
      <ReduxProvider>
        <TrackItem track={track} tracks={tracks} />
      </ReduxProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
