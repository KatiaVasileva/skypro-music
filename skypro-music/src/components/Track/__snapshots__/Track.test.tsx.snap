// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Track component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="playlistItem"
  >
    <div
      class="playlistTrack"
    >
      <div
        class="trackTitle"
      >
        <div
          class="imageContainer"
        >
          <div
            class="trackTitleImage"
          >
            <svg
              class="trackTitleSvg"
            >
              <use
                xlink:href="/img/icon/sprite.svg#icon-note"
              />
            </svg>
          </div>
        </div>
        <div>
          <a
            class="trackTitleLink"
            href="#"
          >
            Track1 
            <span
              class="trackTitleSpan"
            />
          </a>
        </div>
      </div>
      <div
        class="trackAuthor"
      >
        <a
          class="trackAuthorLink"
          href="http://"
        >
          Author1
        </a>
      </div>
      <div
        class="trackAlbum"
      >
        <a
          class="trackAlbumLink"
          href="http://"
        >
          Album1
        </a>
      </div>
      <div
        class=""
      >
        <svg
          class="trackLikeSvg"
        >
          <use
            xlink:href="/img/icon/sprite.svg#icon-like"
          />
        </svg>
      </div>
      <div>
        <span
          class="trackTimeText"
        >
          1:40
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
